core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,399 @@
+    long newTime = trustCertFile.lastModified();
+    if (newTime != oldTime) {
+      X509Certificate[] certificates = CertificateUtils.getX509Certificates(
+          new FileInputStream(trustCertFile));
+      updateTrustCredentials(certificates);
+      return newTime;
+    }
+    return 0;
+  }
+
+  // Mainly used to avoid throwing IO Exceptions in java.io.Closeable.
+  public interface Closeable extends java.io.Closeable {
+    @Override public void close();
+  }
+
+  public static Builder newBuilder() {
+    return new Builder().setVerification(Verification.CertificateAndHostNameVerification);
+  }
+
+  // The verification mode when authenticating the peer certificate.
+  public enum Verification {
+    // This is the DEFAULT and RECOMMENDED mode for most applications.
+    // Setting this on the client side will do the certificate and hostname verification, while
+    // setting this on the server side will only do the certificate verification.
+    CertificateAndHostNameVerification,
+    // This SHOULD be chosen only when you know what the implication this will bring, and have a
+    // basic understanding about TLS.
+    // It SHOULD be accompanied with proper additional peer identity checks set through
+    // {@code PeerVerifier}(nit: why this @code not working?). Failing to do so will leave
+    // applications to MITM attack.
+    // Also note that this will only take effect if the underlying SDK implementation invokes
+    // checkClientTrusted/checkServerTrusted with the {@code SSLEngine} parameter while doing
+    // verification.
+    // Setting this on either side will only do the certificate verification.
+    CertificateOnlyVerification,
+    // Setting is very DANGEROUS. Please try to avoid this in a real production environment, unless
+    // you are a super advanced user intended to re-implement the whole verification logic on your
+    // own. A secure verification might include:
+    // 1. proper verification on the peer certificate chain
+    // 2. proper checks on the identity of the peer certificate
+    InsecurelySkipAllVerification,
+  }
+
+  // Additional custom peer verification check.
+  // It will be used when checkClientTrusted/checkServerTrusted is called without
+  // {@code Socket}/{@code SSLEngine} parameter.
+  public interface PeerVerifier {
+    /**
+     * Verifies the peer certificate chain. For more information, please refer to
+     * {@code X509ExtendedTrustManager}.
+     *
+     * @param  peerCertChain  the certificate chain sent from the peer
+     * @param  authType the key exchange algorithm used, e.g. "RSA", "DHE_DSS", etc
+     */
+    void verifyPeerCertificate(X509Certificate[] peerCertChain, String authType)
+        throws CertificateException;
+  }
+
+  // Additional custom peer verification check.
+  // It will be used when checkClientTrusted/checkServerTrusted is called with the {@code Socket}
+  // parameter.
+  public interface SslSocketPeerVerifier {
```

Comment:
This and the SslEngine version should be part of the same interface.

If this is specified but PeerVerifier is not, and `socket == null && engine == null` then it seems we should fail the trust check.